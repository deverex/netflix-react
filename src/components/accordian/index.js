import React, { useContext, createContext, useState } from "react";
import {
  Container,
  Inner,
  Title,
  Item,
  Header,
  Body,
} from "./styles/accordian";

const ToggleContext = createContext();

const Accordian = ({ children, ...restProps }) => {
  return (
    <Container {...restProps}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const AccordianTitle = ({ children, ...restProps }) => {
  return <Title {...restProps}>{children}</Title>;
};

const AccordianItem = ({ children, ...restProps }) => {
  const [toggleShow, setToggleShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...restProps}>{children}</Item>
    </ToggleContext.Provider>
  );
};

const AccordianHeader = ({ children, ...restProps }) => {
  const { toggleShow, setToggleShow } = useContext(ToggleContext);
  return (
    <Header
      onClick={() => setToggleShow((toggleShow) => !toggleShow)}
      {...restProps}
    >
      {children}
      {toggleShow ? (
        <img src="/images/icons/close-slim.png" alt="Close" />
      ) : (
        <img src="/images/icons/add.png" alt="Open" />
      )}
    </Header>
  );
};

const AccordianBody = ({ children, ...restProps }) => {
  const { toggleShow } = useContext(ToggleContext);
  return toggleShow ? <Body {...restProps}>{children}</Body> : null;
};

Accordian.Title = AccordianTitle;
Accordian.Item = AccordianItem;
Accordian.Header = AccordianHeader;
Accordian.Body = AccordianBody;

export default Accordian;
