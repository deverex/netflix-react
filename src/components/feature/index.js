import React from "react";
import { Container, Title, SubTitle } from "./styles/feature";

const Feature = ({ children, ...restProps }) => {
  return <Container {...restProps}>{children}</Container>;
};

const FeatureTitle = ({ children, ...restProps }) => {
  return <Title {...restProps}>{children}</Title>;
};

const FeatureSubTitle = ({ children, ...restProps }) => {
  return <SubTitle {...restProps}>{children}</SubTitle>;
};

Feature.Title = FeatureTitle;
Feature.SubTitle = FeatureSubTitle;

export default Feature;
