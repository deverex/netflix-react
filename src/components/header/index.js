import React from "react";
import { Link as ReactRouterLink } from "react-router-dom";

import {
  Background,
  Container,
  Logo,
  ButtonLink,
  Feature,
  FeatureCallOut,
  Text,
  Link,
  Group,
  Picture,
  Profile,
  Dropdown,
} from "./styles/header";

const Header = ({ bg = true, children, ...restProps }) => {
  return bg ? <Background {...restProps}>{children}</Background> : children;
};

const HeaderPicture = ({ src, ...restProps }) => {
  return <Picture {...restProps} src={`/images/users/${src}.png`} />;
};

const HeaderProfile = ({ children, ...restProps }) => {
  return <Profile {...restProps}>{children}</Profile>;
};

const HeaderDropdown = ({ children, ...restProps }) => {
  return <Dropdown {...restProps}>{children}</Dropdown>;
};

const HeaderFeature = ({ children, ...restProps }) => {
  return <Feature {...restProps}>{children}</Feature>;
};

const HeaderFeatureCallOut = ({ children, ...restProps }) => {
  return <FeatureCallOut {...restProps}>{children}</FeatureCallOut>;
};

const HeaderFrame = ({ children, ...restProps }) => {
  return <Container {...restProps}>{children}</Container>;
};

const HeaderGroup = ({ children, ...restProps }) => {
  return <Group {...restProps}>{children}</Group>;
};

const HeaderLogo = ({ to, children, ...restProps }) => {
  return (
    <ReactRouterLink to={to}>
      <Logo {...restProps} />
    </ReactRouterLink>
  );
};

const HeaderButtonLink = ({ children, ...restProps }) => {
  return <ButtonLink {...restProps}>{children}</ButtonLink>;
};

const HeaderText = ({ children, ...restProps }) => {
  return <Text {...restProps}>{children}</Text>;
};

const HeaderTextLink = ({ children, ...restProps }) => {
  return <Link {...restProps}>{children}</Link>;
};

Header.Frame = HeaderFrame;
Header.Logo = HeaderLogo;
Header.ButtonLink = HeaderButtonLink;
Header.Feature = HeaderFeature;
Header.FeatureCallOut = HeaderFeatureCallOut;
Header.Text = HeaderText;
Header.TextLink = HeaderTextLink;
Header.Group = HeaderGroup;
Header.Picture = HeaderPicture;
Header.Profile = HeaderProfile;
Header.Dropdown = HeaderDropdown;

export default Header;
