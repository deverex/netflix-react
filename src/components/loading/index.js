import React from "react";
import { Spinner, LockBody, Picture, ReleaseBody } from "./styles/loading";

const Loading = ({ src, ...restProps }) => {
  return (
    <Spinner {...restProps}>
      <LockBody />
      <Picture src={`/images/users/${src}.png`} />
    </Spinner>
  );
};

const LoadingReleaseBody = () => {
  return <ReleaseBody />;
};

Loading.ReleaseBody = LoadingReleaseBody;

export default Loading;
