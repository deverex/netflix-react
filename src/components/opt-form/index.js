import React from "react";
import { Container, Input, Button, Text, Break } from "./styles/opt-form";

const OptForm = ({ children, ...restProps }) => {
  return <Container {...restProps}>{children}</Container>;
};

const OptFormInput = ({ ...restProps }) => {
  return <Input {...restProps} />;
};

const OptFormButton = ({ children, ...restProps }) => {
  return (
    <Button {...restProps}>
      {children}
      <img src="/images/icons/chevron-right.png" alt="Try Now" />
    </Button>
  );
};

const OptFormText = ({ children, ...restProps }) => {
  return <Text {...restProps}>{children}</Text>;
};

const OptFormBreak = ({ children, ...restProps }) => {
  return <Break {...restProps} />;
};

OptForm.Button = OptFormButton;
OptForm.Input = OptFormInput;
OptForm.Text = OptFormText;
OptForm.Break = OptFormBreak;

export default OptForm;
