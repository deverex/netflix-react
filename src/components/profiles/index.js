import React from "react";
import { Container, Title, Item, List, Picture, Name } from "./styles/profiles";

const Profiles = ({ children, ...restProps }) => {
  return <Container {...restProps}>{children}</Container>;
};

const ProfilesTitle = ({ children, ...restProps }) => {
  return <Title {...restProps}>{children}</Title>;
};

const ProfilesList = ({ children, ...restProps }) => {
  return <List {...restProps}>{children}</List>;
};

const ProfilesItem = ({ children, ...restProps }) => {
  return <Item {...restProps}>{children}</Item>;
};

const ProfilesUser = ({ children, ...restProps }) => {
  return <Item {...restProps}>{children}</Item>;
};

const ProfilesPicture = ({ children, src, ...restProps }) => {
  return (
    <Picture
      {...restProps}
      src={src ? `/images/users/${src}.png` : "/images/misc/loading.gif"}
    />
  );
};

const ProfilesName = ({ children, ...restProps }) => {
  return <Name {...restProps}>{children}</Name>;
};

Profiles.Title = ProfilesTitle;
Profiles.List = ProfilesList;
Profiles.Item = ProfilesItem;
Profiles.User = ProfilesUser;
Profiles.Picture = ProfilesPicture;
Profiles.Name = ProfilesName;

export default Profiles;
