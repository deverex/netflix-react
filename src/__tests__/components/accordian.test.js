import React from "react";
import { render, fireEvent, queryByText } from "@testing-library/react";
import faqsData from "../../fixtures/faqs.json";
import { Accordian } from "../../components";

describe("<Accordian />", () => {
  it("renders the <Accordian /> with populated data", () => {
    const { container, getByText } = render(
      <Accordian>
        <Accordian.Title>Frequently Asked Questions</Accordian.Title>
        {faqsData.map((item) => (
          <Accordian.Item key={item.id}>
            <Accordian.Header>{item.header}</Accordian.Header>
            <Accordian.Body>{item.body}</Accordian.Body>
          </Accordian.Item>
        ))}
      </Accordian>
    );

    expect(getByText("Frequently Asked Questions")).toBeTruthy();
    expect(getByText("What is Netflix?")).toBeTruthy();
    expect(getByText("How much does Netflix cost?")).toBeTruthy();
    expect(getByText("Where can I watch?")).toBeTruthy();
    expect(getByText("How do I cancel?")).toBeTruthy();
    expect(getByText("What can I watch on Netflix?")).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("opens and closes the <Accordian /> component", () => {
    const { container, queryByText } = render(
      <Accordian>
        <Accordian.Title>Frequently Asked Questions</Accordian.Title>
        {faqsData.map((item) => (
          <Accordian.Item key={item.id}>
            <Accordian.Header>{item.header}</Accordian.Header>
            <Accordian.Body>{item.body}</Accordian.Body>
          </Accordian.Item>
        ))}
      </Accordian>
    );

    const whatIsNetflixBodyText =
      "Netflix is a streaming service that offers a wide variety of award-winning TV programmes, films, anime, documentaries and more – on thousands of internet-connected devices. You can watch as much as you want, whenever you want, without a single advert – all for one low monthly price. There's always something new to discover, and new TV programmes and films are added every week!";

    expect(queryByText(whatIsNetflixBodyText)).toBeFalsy();
    fireEvent.click(queryByText("What is Netflix?"));
    expect(queryByText(whatIsNetflixBodyText)).toBeTruthy();

    fireEvent.click(queryByText("What is Netflix?"));
    expect(queryByText(whatIsNetflixBodyText)).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
