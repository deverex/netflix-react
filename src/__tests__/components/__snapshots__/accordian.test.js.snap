// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordian /> opens and closes the <Accordian /> component 1`] = `
<div
  class="accordian__Container-sc-1njbxf3-0 dzSVKY"
>
  <div
    class="accordian__Inner-sc-1njbxf3-1 cNPAgL"
  >
    <h1
      class="accordian__Title-sc-1njbxf3-2 iSAAVC"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        What is Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        How much does Netflix cost?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        Where can I watch?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        How do I cancel?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        What can I watch on Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Accordian /> renders the <Accordian /> with populated data 1`] = `
<div
  class="accordian__Container-sc-1njbxf3-0 dzSVKY"
>
  <div
    class="accordian__Inner-sc-1njbxf3-1 cNPAgL"
  >
    <h1
      class="accordian__Title-sc-1njbxf3-2 iSAAVC"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        What is Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        How much does Netflix cost?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        Where can I watch?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        How do I cancel?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordian__Item-sc-1njbxf3-3 jQpIRi"
    >
      <div
        class="accordian__Header-sc-1njbxf3-4 wWcoE"
      >
        What can I watch on Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;
